x-pack/plugin/ilm/qa/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java

```diff
@@ -223,6 +228,40 @@ public void testRolloverActionWithIndexingComplete() throws Exception {
             getOnlyIndexSettings(client(), originalIndex).get(LifecycleSettings.LIFECYCLE_INDEXING_COMPLETE)));
     }
 
+    public void testRolloverActionWithMaxSinglePrimarySize() throws Exception {
+        String originalIndex = index + "-000001";
+        String secondIndex = index + "-000002";
+        createIndexWithSettings(client(), originalIndex, alias, Settings.builder()
+            .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 3)
+            .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+            .put(RolloverAction.LIFECYCLE_ROLLOVER_ALIAS, alias));
+
+        int totalDocs = randomIntBetween(16, 32);
+        for (int i = 0; i < totalDocs; i++) {
+            index(client(), originalIndex, String.valueOf(i), "foo", "bar_" + i);
+        }
+        refresh(client(), originalIndex);
+
+        List<Integer> primaryShardSizes = getPrimaryShardSizes(client(), originalIndex);
+        Integer largest = Collections.max(primaryShardSizes);
+
+        // create policy
+        createNewSingletonPolicy(client(), policy, "hot", new RolloverAction(
+            null,
+            // the largest size we have from above is index stats size, but that's larger
+            // than the docstats size that we actually evaluate against inside the guts
+            // of rollover, 2048 is a slop factor
+            ByteSizeValue.ofBytes(Long.max(1, largest - 2048)),
+            null, null));
+        // update policy on index
+        updatePolicy(client(), originalIndex, policy);
+
+        assertBusy(() -> assertTrue(indexExists(secondIndex)));
+        assertBusy(() -> assertTrue(indexExists(originalIndex)));
```

Comment:
due to various flakiness issues we usually put a 30 seconds timeout on these as CI might be slow (it defaults to 10 seconds I think)

Reply:
2ad066c9eb7cbd6dd922dcf68854fea61c5b933a